src/vm/i386/umthunkstub.S

```diff
@@ -25,8 +26,153 @@ NESTED_ENTRY TheUMEntryPrestub, _TEXT, UnhandledExceptionHandlerUnix
     jmp     eax     // Tail Jmp
 NESTED_END TheUMEntryPrestub, _TEXT
 
-
+//
+// eax: UMEntryThunk*
+//
 NESTED_ENTRY UMThunkStub, _TEXT, UnhandledExceptionHandlerUnix
-    int     3       // implement here
+
+#define UMThunkStub_LOCALVARS               (3*4)   // UMEntryThunk*, Thread*, Stub*
+#define UMThunkStub_UMENTRYTHUNK_OFFSET     0x04
+#define UMThunkStub_THREAD_OFFSET           0x08
+#define UMThunkStub_PILSTUB_OFFSET          0x0c
+#define UMThunkStub_INT_ARG_SPILL           (5*4)   // for save ecx, edx, ebx, esi, edi
+#define UMThunkStub_FIXEDALLOCSIZE          (UMThunkStub_LOCALVARS+UMThunkStub_INT_ARG_SPILL+4)
+#define UMThunkStub_INT_ARG_OFFSET          (UMThunkStub_LOCALVARS+4)
+
+// return address                           <-- entry ESP
+// saved ebp                                <-- EBP
+// UMEntryThunk*
+// Thread*
+// Stub*
+// save ecx
+// save edx
+// save ebx
+// save esi
+// save edi
+// {optional stack args passed to callee}   <-- new esp
+
+    PROLOG_BEG
+    PROLOG_END
+    sub     esp, UMThunkStub_FIXEDALLOCSIZE
+
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET], ecx
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 0x04], edx
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 0x08], ebx
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 0x0c], esi
```

Comment:
@seanshpark I didn't mean it this way. The ebx, esi and edi are not argument registers. They are callee saved registers. I meant just pushing them in the prolog and popping in the epilog. Otherwise the stack unwinding would not work properly.